### PR TITLE
[TypeScript] Fix onSuccess / onFailure types

### LINKED ISSUE #5761: Create prop onSuccess has wrong typing
According to the [CreateProps typings](https://github.com/marmelab/react-admin/blob/872480a9a4a8491066eb8b1f0f651e48f47e9080/packages/ra-ui-materialui/src/types.ts#L50), the onSuccess callback has the type `(data: Record) => void`.

However, it seems that the true implementation in the [useCreateController](https://github.com/marmelab/react-admin/blob/872480a9a4a8491066eb8b1f0f651e48f47e9080/packages/ra-core/src/controller/details/useCreateController.ts#L167) the type `({ data } : { data: Record}) => void` is used.

Right now the two are incompatible. I would suggest to change the type as this breaks no implementation based on the current behaviour?

### PR DESCRIPTION
**Pull request by fzaninotto**
Closes #5761
Supersedes #5787

### CODE REVIEW

File: packages/ra-core/src/types.ts [hunk 1]
```diff
@@ -286,6 +286,8 @@ export type DataProviderProxy = {
 };
 
 export type MutationMode = 'pessimistic' | 'optimistic' | 'undoable';
+export type OnSuccess = (response?: any) => void;
+export type OnFailure = (error?: any) => void;
```
Comment (djhi) [inline packages/ra-core/src/types.ts:290]: We should probably avoid `any` in favor of `unknow`. See #5834

Comment (fzaninotto) [reply packages/ra-core/src/types.ts:290]: I'm not sure I understand the benefits and shortcomings of `any` vs `unknown`. What I've read at https://stackoverflow.com/questions/51439843/unknown-vs-any tells me that there may be drawbacks for end users (like the obligation to perform type checking on an unknown value before using it).

Let's keep the two discussions separate.

**Review by djhi — CHANGES_REQUESTED** [review]
There is an eslint warning ;)

**Comment by HendrikRoehm** [comment]
Additionally, the main fix for #5761 is missing:
`OnSuccess` should be typed as `({ data } : { data: Record}) => void`

**Comment by djhi** [comment]
@HendrikRoehm:

> Additionally, the main fix for #5761 is missing:

No, `data` is not always a `Record`. It could be an array of `Record` for `useGetList` or anything in the case of custom dataProvider functions

**Comment by HendrikRoehm** [comment]
@djhi I guess the main problem here is that there exists at least 3 different kind of onSuccess properties, which expect different types. Plz check my issue: In the onCreateController, it **need** to be of the type if have mentioned it above. So at least the type there should be changed.

**Comment by fzaninotto** [comment]
@HendrikRoehm You said:

> In the onCreateController, it need to be of the type I have mentioned above

What do you mean? Do you have a compilation error? If so, with which code? 

Also, it's only in dataProvider.create() that the onsuccess callback can expect a defined response with a data key.

**Review by djhi — APPROVED** [review]

**Pull request merged**